Different field types with the same name break CM EditView

joshuaellis opened the pull request:
<!--
Hello 👋 Thank you for submitting a pull request.

To help us merge your PR, make sure to follow the instructions below:

- Create or update the tests
- Create or update the documentation at https://github.com/strapi/documentation
- Refer to the issue you are closing in the PR description: Fix #issue
- Specify if the PR is ready to be merged or work in progress (by opening a draft PR)

Please ensure you read the Contributing Guide: https://github.com/strapi/strapi/blob/main/CONTRIBUTING.md
-->

### What does it do?

* removes `recursivelyFindPathsBasedOnCondition`
* removes `findLeafByPathAndReplace`
* adds `findAllAndReplace`
* refactors `getPopulateForDZ` to use fragments to populate the dynamic zone components

### Why is it needed?

There are two issues here –

* On the FE we weren't passing the index to the relation preperation function in `INIT_FORM` so we (wrongly) assumed that all fields with the relation name in DZs or Repetable Components were relations, this has been changed so we check the type of schema before replacing the value instead.
* On the BE because of how relations & components work (they need to populated) it was confusing which part of the schema wanted to be a `count` in the case of relations vs `populated` in the case of a component, the refactor now uses fragments for better fine grained choosing.

### How to test it?

* automated tests

### Related issue(s)/PR(s)

* resolves #16048

joshuaellis commented:
@gu-stav wanted to give you visibility on the approach.

joshuaellis commented:
Gonna add some more tests around the `path` and `parent` part that are passed to the predicate & replacement functions but this is good for review.

joshuaellis commented:
Hey @derrickmehaffy, 

This release `0.0.0-f78ba52ea63afa7bd8507d320e458a0939a3f709` when used in your example repo appears to fix the issues you were having, feel free to try it out and let us know if you see any other issues 👍🏼

derrickmehaffy commented:
Thanks so much @joshuaellis <3

gu-stav commented on packages/core/admin/admin/src/content-manager/components/EditViewDataManagerProvider/reducer.js:
```diff
@@ -280,68 +218,51 @@ const reducer = (state, action) =>
+          attributes
+        );
+
+        const findAllDynamicAndRegularComponentsAndReplaceWithTempKey = findAllAndReplace(
```
Are dynamic components = repeatable components? And are regular components non-repeatable components?

It is very nit-picky, but if so I'd suggest to rename this to

```suggestion
        const findComponentsAndReplaceWithTempKey = findAllAndReplace(
```

gu-stav commented on packages/core/admin/admin/src/content-manager/components/EditViewDataManagerProvider/tests/reducer.test.js:
```diff
@@ -1390,6 +1473,7 @@ describe('CONTENT MANAGER | COMPONENTS | EditViewDataManagerProvider | reducer',
               "component": {
                 "__temp_key__": 1,
                 "field1": {
+                  "__temp_key__": 1,
```
Should all these fields have the same `__temp_key__`? Could you run into e.g. keying problems because of that?

gu-stav commented on packages/core/admin/admin/src/content-manager/components/EditViewDataManagerProvider/utils/findAllAndReplace.js:
```diff
@@ -0,0 +1,89 @@
+        }
+
+        if (predicate(value, { path: [...path, key], parent: acc })) {
+          path = [...path, key];
+          acc[key] =
+            typeof replacement === 'function'
+              ? replacement(acc[key], { path, parent: acc })
+              : replacement;
+          path = [];
```
I find it a bit confusing on why you assign path to empty it later on here and in the following two instances.

What do you think about:

```suggestion
          path = ;
            typeof replacement === 'function'
              ? replacement(acc[key], { path: [...path, key], parent: acc })
              : replacement;
          path = [];
```

I'm wondering if this would allow you to get rid of the global `path` variable entirely?

joshuaellis replied:
You need to have the global path because when you dive down one layer (for components / dynamic zones etc) you need that path to be remained although this might be able to be part of the func...

Nice idea, i'll give it a look.

joshuaellis replied:
looks like you were on to something!

gu-stav commented on packages/core/admin/admin/src/content-manager/components/EditViewDataManagerProvider/reducer.js:
```diff
@@ -280,68 +218,51 @@ const reducer = (state, action) =>
-         * Therefore we reset these bits of state to the correct data type
-         * which is an array. Hence why we replace those fields.
-         */
-        const mergeDataWithPreparedRelations = relationalFieldPaths
```
QA will and the tests will probably cover that automatically: where do you handle relations now, which are not part of components? From the code this is not immediately clear to me.

joshuaellis replied:
`findAllRelationsAndReplaceWithEmptyArray` im not sure what line this is on.

gu-stav reviewed (commented):
Very nice work. All the comments are quite minor. 👏🏼  This is a very good evolution of the previous helpers.

I'll do a thorough QA now and let you know once it is finished.

gu-stav commented:
QA:

## Dynamic zones

Following setup: a dynamic-zone contains two components: one 1:1 and one 1:m relation to the same target-content-type (categories) with the same name

The admin API (screenshot) sends a count of 1 for both of them, but the FE neither displays it, not fetches it the relations for the second component. The order of the components doesn't matter - it is always the second that fails.

<img width="649" alt="Screenshot 2023-03-15 at 08 16 18" src="https://user-images.githubusercontent.com/2244375/225234509-9570605e-55b1-4362-9e85-12024956278b.png">

joshuaellis commented:
> The admin API (screenshot) sends a count of 1 for both of them, but the FE neither displays it, not fetches it the relations for the second component. The order of the components doesn't matter - it is always the second that fails.

The API isn't sending me any results it seems... @Marc-Roig might need to look at this...

gu-stav commented:
@joshuaellis Ok but that means it should be unrelated to this PR and could be tackled as a follow-up to merge this one asap, I guess?

Marc-Roig commented:
@gu-stav  we also made changes on the backend for dz population. We should be tackling this in this pr imo

joshuaellis commented:
Actually looking into it a bit more, i think we're not building the correct path inside RelationDataInputManager. the network tab wasn't that clear for me here.

joshuaellis commented:
> The admin API (screenshot) sends a count of 1 for both of them, but the FE neither displays it, not fetches it the relations for the second component. The order of the components doesn't matter - it is always the second that fails.

@gu-stav Solved this issue, the dynamic zones were all getting the same (incorrect tempkey) i've fixed this and added more to the tests to ensure we can catch this next time.

Marc-Roig commented:
Was testing this and can not see anything backend related either

gu-stav commented:
@joshuaellis cool, thanks. I'll test again and let you know once I am finished. Couldn't find anything else so far.

gu-stav reviewed (approved):
Works well for me - great work!

Convly reviewed (approved):
[LGTM](https://github.com/strapi/strapi/pull/16061/files#diff-0dbffbd4cd01b7826ea0a0926c72b04c7f9406134a34fe9b9ee3fa089f57acb0)
